### PR TITLE
Add shared-memory-dict

### PR DESCRIPTION
**Pull request by maxb2**
Checklist
- [x] Title of this PR is meaningful: e.g. "Adding my_nifty_package", not "updated meta.yaml".
- [x] License file is packaged (see [here](https://github.com/conda-forge/staged-recipes/blob/5eddbd7fc9d1502169089da06c3688d9759be978/recipes/example/meta.yaml#L64-L73) for an example).
- [x] Source is from official source.
- [x] Package does not vendor other packages. (If a package uses the source of another package, they should be separate packages or the licenses of all packages need to be packaged).
- [x] If static libraries are linked in, the license of the static library is packaged.
- [x] Package does not ship static libraries. If static libraries are needed, [follow CFEP-18](https://github.com/conda-forge/cfep/blob/main/cfep-18.md).
- [x] Build number is 0.
- [x] A tarball (`url`) rather than a repo (e.g. `git_url`) is used in your recipe (see [here](https://conda-forge.org/docs/maintainer/adding_pkgs.html#build-from-tarballs-not-repos) for more details).
- [x] GitHub users listed in the maintainer section have posted a comment confirming they are willing to be listed there.
- [x] When in trouble, please check our [knowledge base documentation](https://conda-forge.org/docs/maintainer/knowledge_base.html) before pinging a team.

### CODE REVIEW

**Comment by maxb2** [comment]
I'm willing to be listed as a maintainer

**Comment by maxb2** [comment]
@conda-forge/help-python, ready for review!

**Comment by maxb2** [comment]
@conda-forge/staged-recipes, ready for review!

File: recipes/shared-memory-dict/meta.yaml
Comment (jakirkham) [inline recipes/shared-memory-dict/meta.yaml:21]: Need to include `python` in `run`

```suggestion
    - poetry
  run:
    - python
```

File: recipes/shared-memory-dict/meta.yaml
Comment (BastianZim) [inline recipes/shared-memory-dict/meta.yaml:60]: This section needs to be formatted in a different way. Please have a look at the mathplotlib feedstock as an example on how to do it: https://github.com/conda-forge/matplotlib-feedstock/blob/main/recipe/meta.yaml

**Review by BastianZim — CHANGES_REQUESTED** [review]

File: recipes/shared-memory-dict/meta.yaml
Comment (maxb2) [reply recipes/shared-memory-dict/meta.yaml]: See [778a669](https://github.com/conda-forge/staged-recipes/pull/19855/commits/778a66946d8d34b23308b0f2008d9758f1981e48). FYI, I used the new grayskull feature to [automatically add optional dependencies](https://github.com/conda-incubator/grayskull/pull/341) to make the original recipe.

Comment (BastianZim) [reply recipes/shared-memory-dict/meta.yaml:60]: Great! Seems though like there is a bug in Grayskull as the build script needs to be added like this: https://github.com/conda-forge/sagemaker-inference-toolkit-feedstock/blob/main/recipe/build_base.sh

can you also please open an issue in Grayskull so that this gets incorporated?

**Comment by maxb2** [comment]
@BastianZim, I updated the recipe to follow [sagemaker-inference-toolkit-feedstock](https://github.com/conda-forge/sagemaker-inference-toolkit-feedstock/blob/main/recipe/meta.yaml) like you suggested, but the import test in subsequent packages keeps [failing](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=557878&view=logs&j=6f142865-96c3-535c-b7ea-873d86b887bd&t=22b0682d-ab9e-55d7-9c79-49f3c3ba4823&l=1265). I don't know what else to try.

**Comment by BastianZim** [comment]
I'm currently on my phone but I'll have a look in an hour or so.

File: recipes/shared-memory-dict/meta.yaml
Comment (BastianZim) [inline recipes/shared-memory-dict/meta.yaml:77]:
```suggestion
    - maxb2
  feedstock-name: shared-memory-dict
```
Unrelated to the error, this just makes sure that the feedstock name is correct.

File: recipes/shared-memory-dict/meta.yaml
Comment (BastianZim) [inline recipes/shared-memory-dict/meta.yaml:5]:
```suggestion
  name: {{ name|lower }}-meta
```
This package will actually not be built, it is just a meta package.

**Review by BastianZim — COMMENTED** [review]
I can't see the error either but let's make sure that everything is in order and then I can try tracking it down.

File: recipes/shared-memory-dict/meta.yaml
Comment (maxb2) [reply recipes/shared-memory-dict/meta.yaml]: Will do once this gets merged!

File: recipes/shared-memory-dict/meta.yaml
Comment (BastianZim) [inline recipes/shared-memory-dict/meta.yaml:23]:
```suggestion
        - python >=3.8,<4.0
```
Upstream lists 3.8: https://github.com/luizalabs/shared-memory-dict/blob/ff8726973064afaec23eceb09b031abf535a8b4b/pyproject.toml#L21

File: recipes/shared-memory-dict/meta.yaml
Comment (BastianZim) [inline recipes/shared-memory-dict/meta.yaml:52]: To make your life easier, you can also just install `shared-memory-dict-aiocache` and `shared-memory-dict-django` here. Then you only need to update dependencies once.

**Review by BastianZim — COMMENTED** [review]
I have added two small comments.

For the overall problem: I don't know unfortunately what is causing it but one thing I noticed in the [logs](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=558009&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1385) is that pip installs the build tools (poetry etc) since they are not specified here. Upstream doesn't do that either though so I would first recommend asking them to add the build tools to their [pyproject.toml](https://python-poetry.org/docs/pyproject/#poetry-and-pep-517).

You can also add `--no-index` to the build script which will maybe make it fail sooner.

Unfortunately, multi-output scripts are very finicky, so the error might be somewhere else. 

@ocefpaf Have you seen this before and maybe have an idea?

File: recipes/shared-memory-dict/meta.yaml
Comment (maxb2) [reply recipes/shared-memory-dict/meta.yaml:23]: Hmm, I guess `grayskull` missed that.

**Comment by maxb2** [comment]
> For the overall problem: I don't know unfortunately what is causing it but one thing I noticed in the [logs](https://dev.azure.com/conda-forge/feedstock-builds/_build/results?buildId=558009&view=logs&j=e35d4f76-8ff2-5536-d795-df91e63eb9f7&t=fa7b4b17-b6ff-5c9c-8cfc-15f888c92310&l=1385) is that pip installs the build tools (poetry etc) since they are not specified here. Upstream doesn't do that either though so I would first recommend asking them to add the build tools to their [pyproject.toml](https://python-poetry.org/docs/pyproject/#poetry-and-pep-517).

Upstream does have `poetry` set as the [build backend](https://github.com/luizalabs/shared-memory-dict/blob/main/pyproject.toml#L50-L52), but it is using the old way to do it. I'll submit a PR to update it.

**Comment by BastianZim** [comment]
Ahh true — must've missed that. Please add poetry to host then.

File: recipes/shared-memory-dict/meta.yaml
Comment (carterbox) [inline recipes/shared-memory-dict/meta.yaml:26]:
```suggestion
      run:
        - python >=3.8,<4.0
      run_constrained:
        - aiocache >=0.11.1,<0.12.0
        - django >=3.0.8,<4.0.0
```
You actually need to specify these version constraints here and not in the metapackages because if these packages are installed manually, they still need to meet the constraints.

**Review by carterbox — CHANGES_REQUESTED** [review]

**Comment by carterbox** [comment]
Needed to add all of the special pip args that are set automatically in the top package.

**Comment by maxb2** [comment]
Thanks for all the help getting this merged!

**Pull request merged**
